### PR TITLE
[Dynamic Batching] Add world logging capability

### LINKED ISSUE #2912: World logging not working with eval_model.py with dynamic batching
Use this to open other questions or issues, and provide context here.
When I tried `multiprocessing_eval.py --dynamic-batching full`, it gave me this error `AttributeError: 'DynamicBatchWorld' object has no attribute 'acts'`. It seems dynamic batching is not supported in `world_logging`?

### PR DESCRIPTION
**Pull request by emilydinan**
**Patch description**
Currently dynamic batching fails with world logging because it doesn't log `self.acts` in the way that other worlds do. Fixes #2912.  I added a few attributes to track this. 

I'll note that saving `self._task_acts` and `self._obs` is a little redundant and perhaps a waste of memory. The alternative is to remove all extraneous fields (like text_vec, label_vec, etc.) from the acts in `self._obs` that are added post observation before we add to `self.act`, but it's a little hacky. Thoughts @stephenroller ?

**Testing steps**
Wrote a test

### CODE REVIEW

**Comment by stephenroller** [comment]
A test is really crucial for this one imho. I'm nervous about it. There's an integration teacher that has a variable width, and there's the `test_agents/repeat` or something similar that you could use.

**Comment by emilydinan** [comment]
> Hmm, so a few things:
> 
> * you missed line 1260 which would also need this change
> * act and obs are almost certainly the same thing. Our models do in place editing
> 
> Does DialogPartnerWorld.acts work?
> you could define a dynamic property like
> 
> ```
> @property
> def acts(self):
>     return [self.worlds[i].acts for i in self.last_incices]
> ```
> 
> and set self.last_indices on line 1250, with the batch act. That might be cleaner

i don't see why line 1260 needs this change

**Comment by emilydinan** [comment]
resolved the issues @stephenroller, what do you think of this solution?

i will write a test soon

File: parlai/utils/world_logging.py [hunk 1]
```diff
@@ -89,9 +89,10 @@ def _log_batch(self, world):
         batch_act = world.get_acts()
         parleys = zip(*batch_act)
         for i, parley in enumerate(parleys):
-            self._add_msgs(parley, idx=i)
+            idx = parley[0]['dyn_batch_idx'] if 'dyn_batch_idx' in parley[0] else i
```
Comment (stephenroller) [inline parlai/utils/world_logging.py:92]: leave a comment please

**Comment by stephenroller** [comment]
This solution seems good to me. Would just like to see a test and some comments in both DynamicBatchWorld and WorldLogger pointing out the coupling between them.

**Comment by emilydinan** [comment]
> why would we need to remove the post-processing stuff from an obs prior to logging (tensors etc.)?

when we log, we typically don't include that info because we want to see exactly what was handed to the model from the teacher. plus it gets messy/hard to read when you log all of that stuff

File: parlai/utils/world_logging.py [hunk 2]
```diff
@@ -81,15 +81,21 @@ def _add_episode(self, episode):
...
         batch_act = world.get_acts()
         parleys = zip(*batch_act)
         for i, parley in enumerate(parleys):
-            self._add_msgs(parley, idx=i)
+            # in dynamic batching, we only return `batchsize` acts, but the
+            # 'dyn_batch_idx' key in the task act corresponds the episode index
+            # in the buffer
+            idx = parley[0]['dyn_batch_idx'] if 'dyn_batch_idx' in parley[0] else i
+            self._add_msgs(parley, idx=idx)
             if world.worlds[i].episode_done():
```
Comment (klshuster) [inline parlai/utils/world_logging.py:97]: does this need to be `world.worlds[idx]` as well?

File: parlai/core/worlds.py [hunk 3]
```diff
@@ -1245,15 +1253,20 @@ def parley(self):
...
             # we need to make sure that the teachers saw the result
             self.worlds[i].get_task_agent().observe(act)
             # and that the agent copies saw their own voice
             self.worlds[i].get_model_agent().self_observe(act)
-
             # move these worlds forward
             act = self.worlds[i].get_task_agent().act()
+            # we log the task act and the index of the act
+            # in the buffer for world logging purposes
+            self._task_acts[i] = act
```
Comment (klshuster) [inline parlai/core/worlds.py:1267]: sorry, could you explain again why we do this twice in `parley`?

File: parlai/utils/world_logging.py (hunk 2 above)
Comment (emilydinan) [reply parlai/utils/world_logging.py:97]: yes great catch thank you

File: parlai/core/worlds.py (hunk 3 above)
Comment (emilydinan) [reply parlai/core/worlds.py:1267]: ah yes so for the indices we selected, we have to move the subworlds forward and get the next task act from the teacher

Comment (stephenroller) [reply parlai/core/worlds.py:1267]: yeah it's dynamic batching. Dynb doesn't in parley() ends with all subworlds ready for the model's turn.

**Pull request merged**